File: tree_right_side.py

# DFS Approach by recursing on right side
# Time - O(N)
# Space - O(H)

def rightSideView(self, root: TreeNode) -> List[int]:
    
    self.result = []
    
    if root == None:
        return self.result
    
    self.dfs(root, 0)
    return self.result
    
    

def dfs(self, root, level):
    
    # base case
    if root == None:
        return 
    
    # logic
    if level == len(self.result):
        self.result.append(root.val)
        
    self.dfs(root.right, level + 1)
    
    self.dfs(root.left, level + 1)




# DFS Approach by recursing on left side to get right view of tree, we need to update the level at result with current val
# Time - O(N)
# Space - O(H)

def rightSideView(self, root: TreeNode) -> List[int]:
    
    self.result = []
    
    if root == None:
        return self.result
    
    self.dfs(root, 0)
    return self.result
    
    

def dfs(self, root, level):
    
    # base case
    if root == None:
        return 
    
    # logic
    if level == len(self.result):
        self.result.append(root.val)
    
    else:
        # in the case of updating when we take left view to get right view, we update the level with values accordingly
        self.result[level] = root.val
        
    self.dfs(root.left, level + 1)
    
    self.dfs(root.right, level + 1)
    
        
        
        
        
# BFS Solution
# Time - O(N)
# Space - O(N)

def rightSideView(self, root: TreeNode) -> List[int]:

        
    result = []
    if root == None:
        return result
    
    queue = []
    queue.append(root)
    
    while queue != []:
        
        size = len(queue)
        for i in range(size):
            current = queue.pop(0)
            if i == size - 1:
                result.append(current.val)
                
            if current.left != None:
                queue.append(current.left)
                
            if current.right != None:
                queue.append(current.right)
                
    return result
        
        
    

File: treeRightSide.java
//Time Complexity: O(n)
//Space Complexity: O(n)

import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val) {
        this.val = val;
    }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    List<Integer> lis;

    public static void main(String[] args) {
        Solution sol = new Solution();
        TreeNode root = new TreeNode(1);
        TreeNode node1 = new TreeNode(2);
        TreeNode node2 = new TreeNode(3);
        TreeNode node3 = new TreeNode(5);
        TreeNode node4 = new TreeNode(4);
        node1.right = node3;
        node2.right = node4;
        root.left = node1;
        root.right = node2;
        List<Integer> result = sol.treeRightSide(root);
        System.out.println(result);
    }
    public List<Integer> treeRightSide(TreeNode root) {
        if(root == null) {
            return new ArrayList<Integer>();
        }

        Queue<TreeNode> q = new LinkedList<>();
        lis = new ArrayList<>();
        q.add(root);

        bfs(root, q, lis);

        return lis;
    }

    private void bfs(TreeNode root, Queue<TreeNode> q, List<Integer> lis) {

        while(!q.isEmpty()) {
            int size = q.size();

            for(int i = 0; i < size; i++) {
                TreeNode curr = q.poll();

                if(i == size - 1) {
                    lis.add(curr.val);
                }

                if(curr.left != null) {
                    q.add(curr.left);
                }
                if (curr.right != null) {
                    q.add(curr.right);
                }
            }
        }
    }
}
